Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,582 @@
+        /// Logs information about an encountered exception to the backend data store.
+        /// </summary>
+        /// <param name="source">Source of the exception.</param>
+        /// <param name="message">Detailed description of the exception.</param>
+        /// <returns>true if logging was successful, otherwise false.</returns>
+        protected virtual bool LogError(string source, string message)
+        {
+            if (string.IsNullOrWhiteSpace(SettingsCategory) || string.IsNullOrWhiteSpace(source) || string.IsNullOrWhiteSpace(message))
+                return false;
+
+            Log.Publish(MessageLevel.Error, MessageFlags.SecurityMessage, source, message);
+
+            if (!UseDatabaseLogging)
+                return false;
+
+            try
+            {
+                using (AdoDataConnection connection = new AdoDataConnection(SettingsCategory))
+                {
+                    connection.ExecuteNonQuery("INSERT INTO ErrorLog (Source, Message) VALUES ({0}, {1})", source, message);
+                }
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                // Writing data will fail for read-only databases;
+                // all we can do is track last exception in this case
+                LastException = ex;
+                Log.Publish(MessageLevel.Warning, MessageFlags.SecurityMessage, "LogErrorToDatabase", "Failed to log error to database.", "Database or ErrorLog table may be read-only or inaccessible.", ex);
+            }
+            
+            return false;
+        }
+
+        /// <summary>
+        /// Exchange Authorization Code for a Token
+        /// </summary>
+        /// <param name="code"> The Authorization Code returned by the Auth Server</param>
+        private TokenResponse GetToken(string code)
+        {
+            Dictionary<string, string> postParams = new Dictionary<string, string>() {
+                { "grant_type", "authorization_code" },
+                { "client_id", ClientID },
+                { "client_secret", ClientSecret },
+                { "code", code },
+                { "redirect_uri", RedirectURI }
+            };
+
+            void ConfigureRequest(HttpRequestMessage request)
+            {
+                request.RequestUri = new Uri(TokenEndpoint);
+                request.Method = HttpMethod.Post;               
+                request.Content = new FormUrlEncodedContent(postParams);
+            }
+
+            HttpResponseMessage response;
+            using (HttpRequestMessage request = new HttpRequestMessage())
+            {
+                ConfigureRequest(request);
+               
+                response = Client.SendAsync(request).Result;
+                if (!response.IsSuccessStatusCode)
+                    return null;
+                return JsonConvert.DeserializeObject<TokenResponse>(response.Content.ReadAsStringAsync().Result);
+                
+            }
```

Comment:
This should dispose of the `HttpResponseMessage`. Also, I've read that `task.GetAwaiter().GetResult()` is nearly always preferred because it handles exceptions more intuitively.

```suggestion
            using (HttpRequestMessage request = new HttpRequestMessage())
            {
                ConfigureRequest(request);
               
                using (HttpResponseMessage response = Client.SendAsync(request).GetAwaiter().GetResult())
                {
                    if (!response.IsSuccessStatusCode)
                        return null;
                    return JsonConvert.DeserializeObject<TokenResponse>(response.Content.ReadAsStringAsync().GetAwaiter().GetResult());
                }
            }
```

Reply:
Furthermore, it might be a bit more readable, more performant, and more adaptable to future API adjustments if you make this method `async`, use `await` on the tasks, and defer the `task.GetAwaiter().GetResult()` to the caller.

```c#
public override bool Authenticate()
{
    ...
    if (!string.IsNullOrEmpty(UserData.LoginID))
        token = GetTokenAsync(UserData.LoginID).GetAwaiter().GetResult();
    ...
}

private async Task<TokenResponse> GetTokenAsync(string code)
{
    ...
    using (HttpRequestMessage request = new HttpRequestMessage())
    {
        ConfigureRequest(request);
       
        using (HttpResponseMessage response = await Client.SendAsync(request))
        {
            if (!response.IsSuccessStatusCode)
                return null;
            return JsonConvert.DeserializeObject<TokenResponse>(await response.Content.ReadAsStringAsync());
        }
    }
}
```